packages/fuels-core/src/types/param_types.rs

```diff
@@ -84,57 +84,86 @@ impl ParamType {
-                let variants_types = variants.param_types();
-                Self::any_nested_heap_types(chain!(generics, variants_types))
+    pub fn children_need_extra_receipts(&self) -> bool {
+        match self {
+            ParamType::Array(inner, _) | ParamType::Vector(inner) => {
+                inner.needs_extra_data_receipt(false)
             }
-            ParamType::Struct {
-                fields, generics, ..
-            } => Self::any_nested_heap_types(chain!(fields, generics)),
+            ParamType::Struct { fields, .. } => fields
+                .iter()
+                .any(|param_type| param_type.needs_extra_data_receipt(false)),
+            ParamType::Enum { variants, .. } => variants
+                .param_types()
+                .iter()
+                .any(|param_type| param_type.needs_extra_data_receipt(false)),
+            ParamType::Tuple(inner_types) => inner_types
+                .iter()
+                .any(|param_type| param_type.needs_extra_data_receipt(false)),
             _ => false,
         }
     }
 
-    fn any_nested_heap_types<'a>(param_types: impl IntoIterator<Item = &'a ParamType>) -> bool {
-        param_types
-            .into_iter()
-            .any(|param_type| param_type.uses_heap_types())
+    pub fn validate_is_decodable(&self) -> Result<()> {
+        match self {
+            ParamType::Enum { variants, .. } => {
+                let all_param_types = variants.param_types();
+                let grandchildren_need_receipts = all_param_types
+                    .iter()
+                    .any(|child| child.children_need_extra_receipts());
+                if grandchildren_need_receipts {
+                    return Err(error!(
+                        InvalidType,
+                        "Enums currently support only one level deep heap types."
+                    ));
+                }
+
+                let num_of_children_needing_receipts = all_param_types
+                    .iter()
+                    .filter(|param_type| param_type.needs_extra_data_receipt(false))
+                    .count();
+                if num_of_children_needing_receipts > 1 {
+                    Err(error!(
+                        InvalidType,
+                        "Enums currently support only one heap-type variant. Found: \
+                        {num_of_children_needing_receipts}"
+                    ))
+                } else {
+                    Ok(())
+                }
+            }
+            _ if self.children_need_extra_receipts() => Err(error!(
+                InvalidType,
+                "Nested heap types are currently not supported except in Enums."
+            )),
+            _ => Ok(()),
+        }
     }
 
-    pub fn is_vm_heap_type(&self) -> bool {
-        matches!(
-            self,
-            ParamType::Vector(..) | ParamType::Bytes | ParamType::String
-        )
+    pub fn needs_extra_data_receipt(&self, top_level_type: bool) -> bool {
```

Comment:
Small nit here, seeing `needs_extra_data_receipt(true)` makes it seem like a setter. Would be a bit more readable if we go for somethign like `is_extra_receipt_needed(true)` or similar.